extmod/moduselect.c

```diff
@@ -172,7 +172,9 @@ STATIC mp_obj_t select_select(uint n_args, const mp_obj_t *args) {
             mp_map_deinit(&poll_map);
             return mp_obj_new_tuple(3, list_array);
         }
+        MP_THREAD_GIL_EXIT();
         MICROPY_EVENT_POLL_HOOK
+        MP_THREAD_GIL_ENTER();
```

Comment:
Did you try putting the GIL exit/enter calls within the EVENT_POLL_HOOK macro itself?  That's how stmhal's threading does it (see EVENT_POLL_HOOK in stmhal/mpconfigport.h).